### PR TITLE
Save game

### PR DESCRIPTION
**Pull request by joelngwt**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,17 +10,55 @@ To run split screen correctly, you need to modify Unreal Engine. Go to GameViewp
...
+API:
+```cpp
+UFUNCTION(BlueprintCallable, Category="Game")
+virtual bool Cloudy_SaveGameToSlot
+(
+    USaveGame * SaveGameObject,
+    const FString & SlotName,
+    const int32 UserIndex,
+    const int32 PCID, // Player Controller ID of the player you are saving
+    bool IsAutosaved  // Is the game autosaved?
```
Comment (ngsp94) [inline README.md:52]: Manual save by player - not considered autosave

File: Source/CloudySaveManager/Private/CloudySaveManager.cpp [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+static const FString BaseUrl = "http://127.0.0.1:8000";
+static const FString AuthUrl = "/api-token-auth/";
+static const FString SaveDataUrl = "/save-data/";
+static FString Token;
+
+// Automatically starts and shuts down when UE4 is started/closed
+void CloudySaveManagerImpl::StartupModule()
+{
+    UE_LOG(LogTemp, Warning, TEXT("CloudySaveManager started"));
+    AttemptAuthentication(TEXT("joel"), TEXT("1234")); // Token variable will be populated with robot's token
```
Comment (ngsp94) [inline Source/CloudySaveManager/Private/CloudySaveManager.cpp]: Not fixed token. Token changes with every request, so have to authenticate each time.

File: Source/CloudySaveManager/Private/CloudySaveManager.cpp [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+}
+
+bool CloudySaveManagerImpl::UploadFile(FString Filename, int32 PlayerControllerId, bool IsAutosaved)
+{
+    bool RequestSuccess = false;
+
+    CURL *curl;
+    CURLcode res;
+    std::string readBuffer;
+    readBuffer.clear();
```
Comment (ngsp94) [inline Source/CloudySaveManager/Private/CloudySaveManager.cpp:120]: no need to clear just after declaration

File: Source/CloudySaveManager/Private/CloudySaveManager.cpp [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+    return RequestSuccess;
+}
+
+void CloudySaveManagerImpl::OnAuthResponseComplete(FHttpRequestPtr Request, 
+                                                   FHttpResponsePtr Response, bool bWasSuccessful)
+{
+    if (bWasSuccessful)
+    {
+        FString MessageBody;
```
Comment (ngsp94) [inline Source/CloudySaveManager/Private/CloudySaveManager.cpp:217]: MessageBody not being used

File: Source/CloudySaveManager/Private/CloudySaveManager.cpp [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+            // Try and find it, and failing that, load it
+            UClass* SaveGameClass = FindObject<UClass>(ANY_PACKAGE, *SaveGameClassName);
+            if (SaveGameClass == NULL)
+            {
+                SaveGameClass = LoadObject<UClass>(NULL, *SaveGameClassName);
+            }
+
+            // If we have a class, try and load it.
+            if (SaveGameClass != NULL)
```
Comment (ngsp94) [inline Source/CloudySaveManager/Private/CloudySaveManager.cpp:313]: use if-else here for less confusion

**Comment by ngsp94** [comment]
would be good if all c++ modules share constants for communicating with cloudyweb

File: Source/CloudySaveManager/Private/CloudySaveManager.cpp (hunk 5 above)
Comment (joelngwt) [inline Source/CloudySaveManager/Private/CloudySaveManager.cpp:313]: SaveGameClass is potentially updated on line 309. Cannot if-else.

**Pull request merged**